status-im/status-react
# Display contact verification requests and allow users to decline them

fixes #14171 

### Summary

Displays contact (identity) verification requests in the Activity Center and allows the receiver of such requests to decline them.

- [X] Fix reconciliation when notifications are also present in the "All" tab.
- [X] Introduces a new test helper to verify log messages. See `status-im.test-helpers/using-log-test-appender`.

<img alt="identity-verification-example" src="https://user-images.githubusercontent.com/46027/197662739-81b4190f-561e-4817-a518-4fbe78d9c514.png" height="600" />

### Review notes

- The new test utility `status-im.test-helpers/using-log-test-appender` is used in this PR to verify a dead end (a failure event handler without user feedback) at least logs critical information in case users need to report a bug.

- As new notification types are implemented, @smohamedjavid and I agreed to start taking advantage of multimethods to completely decouple different types of notifications. In the future we'll probably find new abstractions or ways to unify them.

- Status-go is under development to add support for notifications in the existing contact verification flows. The branch https://github.com/status-im/status-go/tree/feature/activity-center-read-unread is not yet merged and it's still prone to change before the final merge, but the implementation seems stable enough for us to move forward with the mobile implementation.

#### Platforms

- Android
- iOS

#### Areas that may be impacted

It's still completely isolated from the rest of the app and behind a feature toggle that's disabled by default.

### Steps to test

- Force mobile app to use compiled version from https://github.com/status-im/status-go/tree/feature/activity-center-read-unread.
- Enable `status-im.utils.config/new-activity-center-enabled?`
- Establish mutual contact approval between two contacts.
- Send contact verification from user A to B either by using a running REPL with the example code below, or using the Desktop client.
- Users can send as many contact verification requests as desired.
- Check that the user receiving such challenges can decline them.

```clojure
(rf/reg-event-fx
   :dev/send-contact-verification-request
   (fn [{:keys [_db]} [_ {:keys [contact-id challenge]}]]
     {::json-rpc/call [{:method     "wakuext_sendContactVerificationRequest"
                        :params     [contact-id challenge]
                        :on-success #(do
                                       (tap> {:success %})
                                       nil)
                        :on-error   #(do
                                       (tap> {:error %})
                                       nil)}]}))

(def alice-id "0x04d0")
(def carol-id  "0x0445")

;; Logged in as Alice, dispatch this event:
(rf/dispatch [:dev/send-contact-verification-request
                {:challenge  "Carol, who am I really?"
                 :contact-id carol-id}])
```

status: ready
